packages/engine/Source/Core/CesiumTerrainProvider.js

```diff
@@ -897,11 +897,29 @@ CesiumTerrainProvider.prototype.requestTileGeometry = function (
 
   if (!defined(layerToUse) && unknownAvailability) {
     // Try again when availability data is ready– Otherwise the tile will be marked as failed and never re-requested
-    return availabilityPromise.then(() =>
-      this.requestTileGeometry(x, y, level, request)
-    );
+    return availabilityPromise.then(() => {
+      // handle promise or undefined return
+      return new Promise((resolve, reject) => {
+        // defer execution to the next event loop
+        setTimeout(() => {
+          const originalResult = this.requestTileGeometry(x, y, level, request);
+          if (originalResult && typeof originalResult.then === "function") {
+            originalResult
+              .then((result) => {
+                resolve(result);
+              })
+              .catch((err) => {
+                reject(err);
+              });
+          } else {
+            // undefined means deferred, wait for the promise to resolve
+            return resolve(this.requestTileGeometry(x, y, level, request));
+          }
+        }, 0); // next tick
```

Comment:
I believe this can all be simplified to:

```suggestion
      return new Promise(resolve=> {
        // defer execution to the next event loop
        setTimeout(() => {
          const promise = this.requestTileGeometry(x, y, level, request);
          resolve(promise);
        }, 0); // next tick
```

The `undefined` value can be passed to the `resolve` function directly. If the value of `promise` is a promise that rejects, the top level promise will still reject as well.

Reply:
@ggetz  the first snippet conditionally resolves based on whether `this.requestTileGeometry` returns a promise or not, while the second snippet always resolves with whatever `this.requestTileGeometry` returns.  In my testing this still seems OK, so I accepted it in favor of conciseness!